health-services/ExerciseSampleCompose/app/src/main/java/com/example/exercisesamplecompose/data/ExerciseClientManager.kt

```diff
@@ -215,5 +242,149 @@ sealed class ExerciseMessage {
         ExerciseMessage()
 }
 
+@RequiresPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+@ExperimentalCoroutinesApi
+fun FusedLocationProviderClient.locationUpdates(): Flow<Location> = callbackFlow {
+    val locationRequest =
+        LocationRequest.Builder(10000).setPriority(Priority.PRIORITY_HIGH_ACCURACY).build()
 
+//    val locationCallback = object : LocationCallback() {
+//        override fun onLocationResult(locationResult: LocationResult) {
+//            for (location in locationResult.locations) {
+//                trySend(location)
+//            }
+//        }
+//    }
+//    requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper())
+//    awaitClose { removeLocationUpdates(locationCallback) }
 
+    val locationListener = LocationListener { location -> trySend(location) }
+
+    requestLocationUpdates(locationRequest, locationListener, Looper.getMainLooper()) // is this the right looper?
+    awaitClose { removeLocationUpdates(locationListener) }
+
+    Log.d("qqqqqq", "requested location updates")
+}
+
+@RequiresPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+fun FusedLocationProviderClient.locationUpdates(callback: (l: Location) -> Unit) {
+    val locationRequest =
+        LocationRequest.Builder(10000).setPriority(Priority.PRIORITY_HIGH_ACCURACY).build()
+
+    val locationCallback = object : LocationCallback() {
+        override fun onLocationResult(locationResult: LocationResult) {
+            for (location in locationResult.locations) {
+                Log.d("qqqqq", "got location")
+                callback(location)
+            }
+        }
+    }
+
+    requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper())
+}
+
+@OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
+@RequiresPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+suspend fun ExerciseClient.setUpdateCallback(
+    callback: ExerciseUpdateCallback,
+    ftpClient: FusedLocationProviderClient
+) {
+    Log.d("qqqqqq", "exerciseupdatecallback")
+
+    val locationRequest =
+        LocationRequest.Builder(10000).setPriority(Priority.PRIORITY_HIGH_ACCURACY).build()
+
+    val locationListener = LocationListener { location -> callback.onExerciseUpdateReceived(
+        exerciseUpdateFromLocation(location))
+    }
+
+    ftpClient.requestLocationUpdates(locationRequest, locationListener, Looper.getMainLooper())
+
+    ftpClient.locationUpdates { location -> callback.onExerciseUpdateReceived(
+        exerciseUpdateFromLocation(location))
+    }
+
+    class Proxy(val obj: ExerciseUpdateCallback) : ExerciseUpdateCallback by obj {
+
+        override fun onExerciseUpdateReceived(update: ExerciseUpdate) {
+
+            val hasLocation = update.latestMetrics.getData(DataType.LOCATION).isNotEmpty()
```

Comment:
Probably worth some comments why you are working around WHS. Or is this something health-services-client should be doing?